### ISSUE TITLE
sudo: effective uid is not 0 [$550]

### ISSUE THREAD

**Issue by MarkSwanson** [nwjs/nw.js#2720]
To provide a simple 'restart wireless' app for my daughter I created a node-webkit app that simply accepted her password and spawned a 'sudo -S /sbin/ifdown wlan0' command (and then ifup...).

This works with node.js from the command line just fine.
However, when I run the exact same code from within node-webkit I get:

stderr:sudo: effective uid is not 0, is /usr/bin/sudo on a file system with the 'nosuid' option set or an NFS file system without root privileges?

I'm not running NFS. The filesystem is plain ext4 and is no using the 'nosuid' option (suid works fine).

Is node-webkit perhaps interfering with sudo on purpose? I shouldn't be in the sandbox; I see:
[30497:1130/200731:ERROR:browser_main_loop.cc(162)] Running without the SUID sandbox! See https://code.google.com/p/chromium/wiki/LinuxSUIDSandboxDevelopment for more information on developing with the sandbox on.

Thoughts?

Thanks!

**Comment by SystemDisc** [comment]
$50 bounty added. Anyone have any ideas why this is happening with sudo?

**Comment by shaynem** [comment]
try this

process.setuid(0);
process.setgid(0);

**Comment by shaynem** [comment]
Even better

if(process.platform == 'darwin' || process.platform == 'linux') {
process.setuid(0);
process.setgid(0);
}

**Comment by SystemDisc** [comment]
To change the UID and GID, doesn't nw.js have to be run as a privileged account? I don't want my users to have to run the app as root. I want the app to elevate via sudo or pkexec when root is necessary and only when necessary. Both sudo and pkexec are effected by this bug. Both are reporting SUID/setUID issues.

**Comment by SystemDisc** [comment]
If I try `process.setuid(0);` or `process.setgid(0); `, I get the following exception thrown at me:
```
{
  "errno": 1,
  "code": "EPERM",
  "syscall": "setuid"
}
```

**Comment by shaynem** [comment]
I haven't tried it without running elevated privs.

Possible to make a wrapper to use gksudo or osascript etc to elevate when
required I'm not sure if that would work or not tho

**Comment by webglider** [comment]
I've come across this issue too while trying to exec a command prefixed with sudo. (v0.12.0)

**Comment by SystemDisc** [comment]
It happens with any program that tries to set uid (pkexec, gksudo, sudo, su, ...)

**Comment by SystemDisc** [comment]
If anyone can take the time to find in which version this bug starts, I can do my best to fix it myself, and create a pull request.

**Comment by amdgigabyte** [comment]
I meet this problem when I try to manage my hosts file from nw

**Comment by swook** [comment]
I may be having a related issue. I'm trying to run a `root`-owned binary which has a `setuid` bit set.

This works with plain io.js / node.js but not with nw.js, which results in the run binary not having root privileges.

**Comment by smolleyes** [comment]
+1 impacted by this bug too try to find the problem but no luck for the moment

**Comment by vilnius-leopold** [comment]
Confrim, I'm having the same/similiar problem when trying to open a privileged desktop app on linux via `spawn('gtk-launch', ['gparted.desktop']);`. gparted should then run polkit to prompt the user for the password but instead I get an error `pkexec must be setuid root` and then the process terminates. This works under nodejs but not in nw.js.

**Comment by SystemDisc** [comment]
Added $500 to bounty

**Comment by rogerwang** [comment]
Fixed it in git. Please try this build.

http://dl.nwjs.io/live-build/04-27-2015/0cea837-4997442-be948af-459755a-2bdc251-1764a45/

**Comment by vilnius-leopold** [comment]
Dude, you're f'n awesome... just tried it, works like a charm! Thanks a lot!

**Comment by SystemDisc** [comment]
Money well spent :P
@rogerwang, when will the fix be available in a stable release, and how do I make sure you get the bounty?

**Comment by rogerwang** [comment]
Thanks @SystemDisc  but I'm not going to claim it because it's my duty to fix the issues here.

I'll try to release 0.12.2 in this week, but the live build is stable with several bug fixes on top of 0.12.1.

**Comment by smolleyes** [comment]
hi

thanks for this fix @rogerwang :)

waiting for the update in the git 

++

**Comment by shaynem** [comment]
+1 thank you :) can now remove my wrappers and not launch as root.

**Comment by smolleyes** [comment]
no way to pull tthis fix in the git before 0.12.2 @rogerwang ? need an urgent update (and i do custom builds)

**Comment by SystemDisc** [comment]
@smolleyes, the nw12 branch contains the fixes.

**Comment by smolleyes** [comment]
humm can t see this, trying to rebuild now but have a link for the pull ?

**Comment by smolleyes** [comment]
just rebuilded form nw12

sudo: le uid effectif n'est pas 0. Est-ce que /usr/bin/sudo est sur un système de fichiers avec l'option « nosuid » ou un système de fichiers NFS sans privilèges root ?

... same problem

**Comment by rogerwang** [comment]
@smolleyes what's the tip commit id of your chromium repo?

**Comment by smolleyes** [comment]
hi roger

don t have a chromium repo why ?

i have it in my .gclient 

 "url"         : "https://github.com/nwjs/chromium.src.git@origin/nw12"

do i need to change it to get this fix ?

thanks

**Comment by rogerwang** [comment]
No you don't need to change it. 

I think your repo is not synced successfully. You can get the commit id in your local directory and compare it with https://github.com/nwjs/chromium.src/commits/nw12

**Comment by smolleyes** [comment]
humm seems right

git rev-parse HEAD
4997442d98d80e4cacb1d81ce90d0a45ccd4b185

and nw report same id

commit 4997442d98d80e4cacb1d81ce90d0a45ccd4b185

i ll search how to override all changes in made in the repo

**Comment by smolleyes** [comment]
just made a git reset --hard and gclient sync
œ
hw can i directly verify if i have the sudo fix ? can t find a push with this title... :p

**Comment by smolleyes** [comment]
hummm

had a small problem ...

```git checkout nw12
La position précédente de HEAD était 4997442... Merge remote-tracking branch 'origin/nw12' into nw12
Basculement sur la branche 'nw12'
Votre branche est en retard sur 'origin/nw12' de 4 commits, et peut être mise à jour en avance rapide.
  (utilisez "git pull" pour mettre à jour votre branche locale)
smo@smo-ubu:~/Documents/nw-build/node-webkit/src$ git pull
Mise à jour 80ebf09..4997442
Fast-forward
 base/process/launch_posix.cc           |  2 ++
 components/crash/app/breakpad_linux.cc |  9 +++++++--
 components/crash/app/breakpad_mac.mm   | 20 ++++++++++++++++++++
 3 files changed, 29 insertions(+), 2 deletions(-)
```

think i t s ok now

**Comment by smolleyes** [comment]
rebuild and same error... strange 

i do a full rebuild...

**Comment by smolleyes** [comment]
ok it works with full rebuild @rogerwang 

thanks :)

ps: still some errors or warnings when closing the app

[27102:0428/134123:ERROR:channel.cc(305)] RawChannel read error (connection broken)

++

**Comment by webglider** [comment]
Thanks for the fix @rogerwang !

- Midhul Varma
<midhul.v@gmail.com>

**Comment by webglider** [comment]
Thanks for the fix @rogerwang !

**Comment by jacob-pan** [comment]
Hello, everyone, shall I ask a question about how to use the "sudo" permission in nw.js? I use follow scripts in nw.js:

```
var exec = require('child_process').exec;
exec('sudo ./pcap', function(err, stdout, stderr) {
        console.log(err);
        console.log(stdout);
        console.log(stderr);
});
```
where `./pcap` is a program which has to use the `sudo` permission.

However, I got a error as `sudo: no tty present and no askpass program specified` in log. How can I do? How can I specified `the askpass program` in OS X? (ps: I'm not using Terminal to launch my app. So I can use `sudo ....`).

Thanks a lot!
